From 70d41d0734f9c55275750d64a5c539476e75dcd2 Mon Sep 17 00:00:00 2001
From: Noam Lessner <nlessner@habana.ai>
Date: Tue, 04 Feb 2020 19:30:02 +0200
Subject: [PATCH] [SW-8229] add recovery and validation flags

When we reboot becuse of fw upgrade or PERST we want to
make sure PCIe link will come up as quick as possiable.
If image failed validation or was in recovery and a warm
reset occurred we run pcie_recovery and only then try to
validate fw again. If not in recovery but fw is already
validated we dont need to do it again, because there won't
be time for it in warm reset.

Change-Id: I02c195caee39e318c74ed86ffa11bcdfcd3b8c1a
Signed-off-by: Noam Lessner <nlessner@habana.ai>
---

diff --git a/app/arcboot_rom.c b/app/arcboot_rom.c
index 8833382..586909c 100644
--- a/app/arcboot_rom.c
+++ b/app/arcboot_rom.c
@@ -16,6 +16,8 @@
 
 struct arcboot_global_s arcboot_rom_global = {
 	.config_fuse = 0,
+	.recovery_flag = 0,
+	.validation_pass = 0
 };
 
 #define ARCBOOT_BRINGUP_ERROR		-1
@@ -23,16 +25,70 @@
 
 static void arcboot_rom_init(void)
 {
-	uint32_t rst_cfg;
+	uint32_t reg;
 
 	/* on warm reset change default to reset ARC and PSOC */
-	rst_cfg = regmap_readl(mmPSOC_GLOBAL_CONF_PRSTN_RST_CFG_L);
-	rst_cfg |= PSOC_GLOBAL_CONF_PRSTN_RST_CFG_L_ARC_MASK;
-	rst_cfg |= PSOC_GLOBAL_CONF_PRSTN_RST_CFG_L_PSOC_MASK;
-	regmap_writel(mmPSOC_GLOBAL_CONF_PRSTN_RST_CFG_L, rst_cfg);
+	reg = regmap_readl(mmPSOC_GLOBAL_CONF_PRSTN_RST_CFG_L);
+	reg |= PSOC_GLOBAL_CONF_PRSTN_RST_CFG_L_ARC_MASK;
+	reg |= PSOC_GLOBAL_CONF_PRSTN_RST_CFG_L_PSOC_MASK;
+	regmap_writel(mmPSOC_GLOBAL_CONF_PRSTN_RST_CFG_L, reg);
+
+	/* read from cold reset flops the recovery and validation flag */
+	reg = regmap_readl(mmPSOC_GLOBAL_CONF_COLD_RST_FLOPS_2);
+	arcboot_rom_global.recovery_flag =
+		(reg & COLD_RST_FLOPS_RECOVERY_FLAG_MASK) >>
+		 COLD_RST_FLOPS_RECOVERY_FLAG_SHIFT;
+
+	arcboot_rom_global.validation_pass  =
+		(reg & COLD_RST_FLOPS_VALIDATION_PASS_MASK) >>
+		 COLD_RST_FLOPS_VALIDATION_PASS_SHIFT;
 
 	regmap_writel(SPI_FLASH_BAUDR_DIVISOR, mmPSOC_QSPI_BAUDR);
 	regmap_writel(write_en_16MB, mmPSOC_SECURITY_SPI_WRITE_LIMIT);
+}
+
+static void arcboot_rom_in_recovery(void)
+{
+	/* if recovery flag is set, bringup PCIe in recovery mode */
+	if (arcboot_rom_global.recovery_flag)
+		pcie_recovery();
+}
+
+static uint8_t arcboot_rom_fw_is_not_validated(void)
+{
+	/* return if fw is already validated */
+	return (!arcboot_rom_global.validation_pass);
+}
+
+static void arcboot_rom_validation_fail(void)
+{
+	uint32_t reg;
+
+	/* validation failed, set recovery flag and clear validation */
+	reg = regmap_readl(mmPSOC_GLOBAL_CONF_COLD_RST_FLOPS_2);
+	reg |= COLD_RST_FLOPS_RECOVERY_FLAG_MASK;
+	reg &= ~COLD_RST_FLOPS_VALIDATION_PASS_MASK;
+	regmap_writel(reg, mmPSOC_GLOBAL_CONF_COLD_RST_FLOPS_2);
+}
+
+static void arcboot_rom_validation_success(void)
+{
+	uint32_t reg;
+
+	/* validation succeeded, set only validation pass*/
+	reg = regmap_readl(mmPSOC_GLOBAL_CONF_COLD_RST_FLOPS_2);
+	reg |= COLD_RST_FLOPS_VALIDATION_PASS_MASK;
+	regmap_writel(reg, mmPSOC_GLOBAL_CONF_COLD_RST_FLOPS_2);
+}
+
+static void arcboot_rom_clear_recovery_flag(void)
+{
+	uint32_t reg;
+
+	/* validation succeeded, set only validation pass*/
+	reg = regmap_readl(mmPSOC_GLOBAL_CONF_COLD_RST_FLOPS_2);
+	reg &= ~COLD_RST_FLOPS_RECOVERY_FLAG_MASK;
+	regmap_writel(reg, mmPSOC_GLOBAL_CONF_COLD_RST_FLOPS_2);
 }
 
 static void arcboot_rom_fuse(void)
@@ -86,7 +142,6 @@
 	uint16_t pcie_fw_size;
 	uint64_t pcie_fw_offset;
 	uint64_t pcie_fw_dest;
-	void (*ppboot)();
 
 	flash_image_offset = arcboot_flash_image_offset(image_ptr);
 	flash_ppboot_pcie_offset = (uint32_t)(FLASH_64KB_PADD +
@@ -125,23 +180,33 @@
 	mem_cpy(pcie_fw_size, pcie_fw_offset, pcie_fw_dest);
 	mem_cpy_wait_finish();
 
-	/* cast ppboot location as function and execute it */
-	ppboot = (void *)(HL_FWSRAM_BASE + PPBOOT_IMAGE_CODE_OFFSET);
-	ppboot();
+	/* we have a valid signed image in fw_sram */
+	arcboot_rom_validation_success();
 
 	return ARCBOOT_BRINGUP_OK;
 }
 
 static void arcboot_rom_bringup_pcie(void)
 {
-	if (arcboot_rom_bringup_pcie_images(ARCBOOT_PRIMARY) !=
-	    ARCBOOT_BRINGUP_OK) {
-		if (arcboot_rom_bringup_pcie_images(ARCBOOT_SECONDARY) !=
-		    ARCBOOT_BRINGUP_OK) {
-			pcie_recovery();
-			/* TODO: deal with PCI */
+	void (*ppboot)();
+
+	if (arcboot_rom_fw_is_not_validated())
+		if (arcboot_rom_bringup_pcie_images(
+		    ARCBOOT_PRIMARY) != ARCBOOT_BRINGUP_OK) {
+			if (arcboot_rom_bringup_pcie_images(
+			    ARCBOOT_SECONDARY) != ARCBOOT_BRINGUP_OK) {
+				arcboot_rom_validation_fail();
+				pcie_recovery();
+				/* PCIe link is up, no valid fw. stay here */
+				while (1)
+					;
+			}
 		}
-	}
+	/* cast ppboot location as function and execute it */
+	ppboot = (void *)(HL_FWSRAM_BASE + PPBOOT_IMAGE_CODE_OFFSET);
+	ppboot();
+
+	arcboot_rom_clear_recovery_flag();
 }
 
 static int arcboot_rom_bringup_ram_images(uint8_t image_ptr)
@@ -185,6 +250,7 @@
 int arcboot_rom_main(void)
 {
 	arcboot_rom_init();
+	arcboot_rom_in_recovery();
 	arcboot_rom_fuse();
 	arcboot_rom_bringup_pcie();
 	arcboot_rom_bringup_arcboot_ram();
diff --git a/common/pcie.c b/common/pcie.c
index e888c05..9103845 100644
--- a/common/pcie.c
+++ b/common/pcie.c
@@ -91,36 +91,3 @@
 	security_pcie_rr_config(spi_en);
 	pcie_config();
 }
-
-/*
- * Wait for Boot Manager to complete PCIe image execution
- * and return to idle state, so it won't initiate UART access anymore
- */
-static void pcie_sync_btl(void)
-{
-	while (regmap_readl(mmPSOC_GLOBAL_CONF_BTM_FSM))
-		;
-}
-
-void pcie_warm_reset_handle(void)
-{
-	/* loop while prstn to de-assert */
-	pcie_loop_while_prstn(PCIE_PRSTN_DEASSERT);
-
-	/* if prstn is asserted then we're in reset, wait 1uS for bootmngr to
-	 * startup before checking its status
-	 */
-	udelay(BOOTMNG_STARTUP_DELAY_AFTER_PERSTN);
-
-	/* wait for bootmngr to reach idle state */
-	pcie_sync_btl();
-
-	/* rerun the pcie_recovery to bring minimal link up again */
-	pcie_recovery();
-}
-
-/* TODO: move to correct file */
-void fw_reset_handle(void)
-{
-
-}
diff --git a/include/arcboot.h b/include/arcboot.h
index 3be2f0d..6c054a6 100644
--- a/include/arcboot.h
+++ b/include/arcboot.h
@@ -45,8 +45,15 @@
 #define PPBOOT_IMAGE_CODE_ALIGN			4
 #define PPBOOT_IMAGE_CODE_OFFSET		(PPBOOT_IMAGE_OFFSET + \
 						 PPBOOT_IMAGE_CODE_ALIGN)
+#define COLD_RST_FLOPS_RECOVERY_FLAG_SHIFT	0x0
+#define COLD_RST_FLOPS_RECOVERY_FLAG_MASK	0x1
+#define COLD_RST_FLOPS_VALIDATION_PASS_SHIFT	0x1
+#define COLD_RST_FLOPS_VALIDATION_PASS_MASK	0x2
+
 struct arcboot_global_s {
 	uint32_t config_fuse;
+	uint8_t recovery_flag;
+	uint8_t validation_pass;
 };
 
 extern struct arcboot_global_s arcboot_rom_global;
